lib/generators/declare_schema/migration/migrator.rb

```diff
@@ -451,31 +451,68 @@ def change_indexes(model, old_table_name, to_rename)
               end
           end
 
-          drop_indexes = (existing_indexes_without_primary_key - model_indexes_without_primary_key).map do |i|
+          indexes_to_drop = existing_indexes_without_primary_key - model_indexes_without_primary_key
+          indexes_to_add = model_indexes_without_primary_key - existing_indexes_without_primary_key
+
+          renamed_indexes_to_drop = []
+          renamed_indexes_to_add = []
+
+          indexes_to_drop.each { |index_to_drop|
+            renamed_columns = index_to_drop.columns.map do |column|
+              to_rename[column]
+            end
+
+            indexes_to_add.each { |index_to_add|
+              if renamed_columns == index_to_add.columns
+                renamed_indexes_to_drop.append(index_to_drop)
+                renamed_indexes_to_add.append(index_to_add)
+              end
+            }
+          }
+
+          drop_indexes = (indexes_to_drop - renamed_indexes_to_drop).map do |i|
             ::DeclareSchema::SchemaChange::IndexRemove.new(new_table_name, i.columns, unique: i.unique, where: i.where, name: i.name)
           end
 
-          add_indexes = (model_indexes_without_primary_key - existing_indexes_without_primary_key).map do |i|
+          add_indexes = (indexes_to_add - renamed_indexes_to_add).map do |i|
             ::DeclareSchema::SchemaChange::IndexAdd.new(new_table_name, i.columns, unique: i.unique, where: i.where, name: i.name)
           end
 
           # the order is important here - adding a :unique, for instance needs to remove then add
           [Array(change_primary_key) + drop_indexes + add_indexes]
         end
 
-        def change_foreign_key_constraints(model, old_table_name)
+        def change_foreign_key_constraints(model, old_table_name, to_rename)
           ActiveRecord::Base.connection.class.name.match?(/SQLite3Adapter/) and raise ArgumentError, 'SQLite does not support foreign keys'
           ::DeclareSchema.default_generate_foreign_keys or return []
 
           existing_fks = ::DeclareSchema::Model::ForeignKeyDefinition.for_model(model, old_table_name)
           model_fks = model.constraint_specs
 
-          drop_fks = (existing_fks - model_fks).map do |fk|
+          fks_to_drop = existing_fks - model_fks
+          fks_to_add = model_fks - existing_fks
+
+          renamed_fks_to_drop = []
+          renamed_fks_to_add = []
+
+          fks_to_drop.each { |fk_to_drop|
+            fks_to_add.each { |fk_to_add|
+              if fk_to_add.child_table_name == fk_to_drop.child_table_name &&
+                fk_to_add.parent_table_name == fk_to_drop.parent_table_name &&
+                fk_to_add.foreign_key == to_rename[fk_to_drop.foreign_key]
```

Comment:
I have a similar question about `nil` from the hash lookup. Can that happen? Any chance that the `foreign_key` method could return `nil` as well? If both of those happened, I'm guessing we wouldn't want to take the `==` path here.

Reply:
I don't have a conclusive answer to this yet. But from looking through the code a bit, I would hazard a guess that `fk_to_add.foreign_key` being `nil` would indicate a bug. Here is my reasoning:

1. We have `fks_to_add = model_fks - existing_fks`.
2. `model_fks` is generated from [here](https://github.com/Invoca/declare_schema/blob/b8bbf31f0a1df01ac61a3f4ac33de09ff252c394/lib/declare_schema/model/habtm_model_shim.rb#L73-L74). Trying to figure out if `foreign_keys` on that line could ever be `[]` was leading me down a rabbit hole, so this is where I'm missing clarity.
3. However, `existing_fks` seems a bit clearer. We generate it from the regex [here](https://github.com/Invoca/declare_schema/blob/b8bbf31f0a1df01ac61a3f4ac33de09ff252c394/lib/declare_schema/model/foreign_key_definition.rb#L33), which looks like it expects the `foreign_key` to be non-nil.
4. Since we subtract one array from the other, and since we expect the `foreign_key` to be non-nil for the part that we subtract, I'd guess that we expect it to be non-nil for the thing that we subtract it from.

We do have the `.presence` [here](https://github.com/Invoca/declare_schema/blob/b8bbf31f0a1df01ac61a3f4ac33de09ff252c394/lib/declare_schema/model/foreign_key_definition.rb#L15) though, so I'm a bit unsure. It might be worth throwing in something to prevent the `nil == nil` situation here just in case it did ever arise, even as a result of another bug.